### PR TITLE
[ci] fastlane supply job

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Currently, releasing the Expo client APK through the Google Play Console is a manual step in the release process. However, it's easily automated using `fastlane supply`. This saves a bit of time and also allows people other than James, Jesse, and Eric to release client APKs.

# How

- Ran `fastlane supply init --package_name=host.exp.exponent` and committed the resulting metadata folder.
- Added the google play service account credentials to circle secrets.
- Added `gulp supply-validation` which ensures that the current APK version has release notes and has been uploaded to staging (steps which should happen before releasing to the Play Store).
- New `client_android_release_google_play` job requires approval, runs this validation step, and if it succeeds, uses fastlane to release the new client APK to production.

An alternative to adding this CI job is to add a gulp script to do this and run it locally each time instead. This might allow slightly easier error handling and could also reduce the chance of accidental releases (though I think this is already pretty low with the validation we've added and with fastlane), but would mean more manual work. 

# Test Plan

CI job uses the exact same fastlane command that I used locally to release 2.9.2 to Google Play. Did not test releasing via CI but tested various other commands to ensure that `fastlane`/auth and validation is working as expected.

### CODE REVIEW

**Comment by nicknovitski** [comment]
Why tools and not tools-public?

**Comment by esamelson** [comment]
It's an internal task that only Expo-associated people should ever need to run, unlike the tasks in `tools-public` which theoretically might be useful to other people developing with Expo (e.g. tasks to build shell apps).

File: .circleci/config.yml [hunk 1]
```diff
@@ -206,7 +213,30 @@ jobs:
...
+      - restore_cache:
+          key: client-android-apk-{{ .Revision }}
+      - run:
+          name: Validate Supply metadata
+          working_directory: ~/expo/tools
+          command: nix run nixpkgs.nodePackages.gulp-cli --command gulp supply-validate
+      - run:
+          name: Upload APK to Google Play and release to production
+          working_directory: ~/expo/fastlane/android
+          command: nix run nixpkgs.fastlane --command fastlane supply --package_name=host.exp.exponent --apk ~/expo/android/exponent-release.apk --track production --skip_upload_images=true --skip_upload_screenshots=true
```
Comment (nicknovitski) [inline .circleci/config.yml]: The supply command can be added as a lane in the existing Fastfile.  The command here could then be `fastlane android:supply`, which would be easier for us to run locally if we ever need to.

**Comment by nicknovitski** [comment]
That makes sense.

Do we have an internal track for this app?  It seems we'd benefit from continuously releasing successfully built apks to ourselves.

File: .circleci/config.yml (hunk 1 above)
Comment (nicknovitski) [reply .circleci/config.yml]: I suspect that the logic of the supply-validate gulp task could also be written as ruby in that lane, making it harder to release an unvalidated apk.

File: .circleci/config.yml [hunk 2]
```diff
@@ -206,7 +213,30 @@ jobs:
...
           path: ~/.gradle/daemon/4.4
+
+  client_android_release_google_play:
+    executor: js
+    steps:
+      - setup
+      - yarn_install:
+          working_directory: ~/expo/tools
+      - yarn_install:
+          working_directory: ~/expo/tools-public
```
Comment (nicknovitski) [inline .circleci/config.yml]: If the gulp script is in `tools`, I don't think you need to install anything in `tools-public`.

**Review by nicknovitski — APPROVED** [review]

File: fastlane/android/metadata/en-US/full_description.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Quickly get and use apps that are specially made for Expo.
```
Comment (ide) [inline fastlane/android/metadata/en-US/full_description.txt]: We should rewrite this whole thing:

> The Expo client app is the quickest way to get started with Expo.
> 
> Developers: Start writing apps with native views and APIs with just your Android device and your computer. Expo is a free and open-source toolchain for creating apps with JavaScript and the React Native engine. Bring your projects to life with interactive gestures and graphics, and try out Expo.
>
> To learn more visit https://expo.io.

File: tools/supply-validate.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  const versionName = buildGradle.match(/versionName '([\d\.]+)'/)[1];
+  
+  if (!fs.existsSync(`../fastlane/android/metadata/en-US/changelogs/${versionCode}.txt`)) {
+    throw new Error(`Missing changelog file at fastlane/android/metadata/en-US/changelogs/${versionCode}.txt`);
+  }
+
+  Config.api.host = 'staging.exp.host';
+  const versionsStaging = await Versions.versionsAsync();
+  if (versionsStaging.androidVersion !== versionName) {
+    throw new Error(`APK version ${versionName} is not yet uploaded to staging. Please download from the client_android job, test and upload manually.`);
```
Comment (ide) [inline tools/supply-validate.js]:
```suggestion
    throw new Error(`APK version ${versionName} is not yet uploaded to staging. Please download from the client_android CI job, test and upload manually.`);
```

**Review by ide — APPROVED** [review]

File: .circleci/config.yml (hunk 2 above)
Comment (esamelson) [reply .circleci/config.yml]: the gulp script step failed when I didn't do this, saying it couldn't find `@expo/json-file` required by `tools-public/project-versions.js`. I couldn't see any obvious dependency on this from `tools` so this was the easiest fix.

**Comment by esamelson** [comment]
Thanks for the comments. I'll go ahead and try adding the command as a lane in the existing Fastfile instead. I wrote the validation logic in JS because I'm not familiar with Ruby at all and so it's much easier for me to both write and understand what's going on. If you both feel strongly about having it be in Ruby I can try to port it over, though.

@nicknovitski we don't have any internal tracks set up, currently. We can look into it in the future but I don't have any bandwidth to set it up right now. Personally I don't think I would find it very useful, as when I use the client I'm either trying to reproduce an issue on the prod client, or developing something myself and building/iterating locally. I'm not opposed to the idea, however.

**Pull request merged**